linkerd/identity/src/imp/rustls.rs

```diff
@@ -0,0 +1,415 @@
+        // more tested.
+        c.enable_tickets = false;
+
+        Some(TrustAnchors(Arc::new(ClientConfig(c))))
+    }
+
+    pub fn certify(&self, key: Key, crt: Crt) -> Result<CrtKey, InvalidCrt> {
+        let mut client = self.0.as_ref().clone().as_ref().clone();
+
+        // Ensure the certificate is valid for the services we terminate for
+        // TLS. This assumes that server cert validation does the same or
+        // more validation than client cert validation.
+        //
+        // XXX: Rustls currently only provides access to a
+        // `ServerCertVerifier` through
+        // `rustls::ClientConfig::get_verifier()`.
+        //
+        // XXX: Once `rustls::ServerCertVerified` is exposed in Rustls's
+        // safe API, use it to pass proof to CertCertResolver::new....
+        //
+        // TODO: Restrict accepted signature algorithms.
+        static NO_OCSP: &[u8] = &[];
+        client
+            .get_verifier()
+            .verify_server_cert(&client.root_store, &crt.chain, (&crt.id).into(), NO_OCSP)
+            .map_err(InvalidCrt)?;
+
+        let k = SigningKey(key.0);
+        let key = rustls::sign::CertifiedKey::new(crt.chain, Arc::new(Box::new(k)));
+        let resolver = Arc::new(CertResolver(key));
+
+        // Enable client authentication.
+        client.client_auth_cert_resolver = resolver.clone();
+
+        // Ask TLS clients for a certificate and accept any certificate issued
+        // by our trusted CA(s).
+        //
+        // XXX: Rustls's built-in verifiers don't let us tweak things as fully
+        // as we'd like (e.g. controlling the set of trusted signature
+        // algorithms), but they provide good enough defaults for now.
+        // TODO: lock down the verification further.
+        //
+        // TODO: Change Rustls's API to Avoid needing to clone `root_cert_store`.
+        let mut server = rustls::ServerConfig::new(
+            rustls::AllowAnyAnonymousOrAuthenticatedClient::new(client.root_store.clone()),
+        );
+        server.versions = TLS_VERSIONS.to_vec();
+        server.cert_resolver = resolver;
+
+        Ok(CrtKey {
+            id: crt.id,
+            expiry: crt.expiry,
+            client_config: Arc::new(ClientConfig(client)),
+            server_config: Arc::new(ServerConfig(server)),
+        })
+    }
+
+    pub fn client_config(&self) -> Arc<ClientConfig> {
+        self.0.clone()
+    }
+}
+
+impl fmt::Debug for TrustAnchors {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("TrustAnchors")
+            .field(self.0.as_ref())
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct InvalidCrt(rustls::TLSError);
+
+impl fmt::Display for InvalidCrt {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl error::Error for InvalidCrt {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        self.0.source()
+    }
+}
```

Comment:
nit: could use `thiserror` here:
```suggestion
#[derive(Clone, Debug, Error)]
#[error(transparent)]
pub struct InvalidCrt(#[from] rustls::TLSError);
```